rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,219 @@
+two isolation levels that will be supported are `Committed` and `Uncommitted`.
+The default when `scan` is called without explicit options will be `Committed`.
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`  
+    pub async fn scan_with_options(
+        &self,
+        range: &DbRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes the active WAL, memtables waiting to be flushed, and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables.  The latest records will
+have precedence following the same rules that a normal `get` query follows.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
+foundation to unify multiple iterators. Additionally, `MemTableIterator` and 
+`SstIterator` provide ready-to-use capabilities to scan tables which may keys
+within the range. 
+
+**Query Isolation**: When using the `Uncommitted` isolation, unflushed
+records recorded in WALs will be used as a source for the key range. Otherwise,
+only committed records in memtables, as well as L0 and SR tables will be used.
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the  query began scanning its range of records. For example,
+a query may include  only a subset of records that were inserted in the same 
+`WriteBatch`. In  other words, range queries will not have snapshot isolation 
+initially. 
+
+[RFC-????](https://github.com/slatedb/slatedb/pull/260) proposes transactional
+semantics for SlateDB. This relies on a monotonic sequence number which is tagged
+on each key update. Range queries could leverage this sequence number to improve
+isolation by filtering any updates with sequence numbers larger than the latest
+at the time the query was executed. We leave this as a gap for future work to address.
+
+### Checkpoint References
+
+We intend to rely on state checkpoints introduced in [RFC-????](https://github.com/slatedb/slatedb/pull/278) in order to prevent the cleanup
+of durable state referenced by the query. When a range scan is created, it will create
+an internal reference to the current checkpoint. This reference will prevent cleanup of
+the checkpoint state until the scan is complete. 
+
+Checkpoints created by `Readers` have an expiry time, which is only refreshed as long
```

Comment:
I'm a smidge confused on this paragraph (I haven't yet read #278). I think you're saying that the client won't heartbeat the checkpoint? Rather, it'll create a checkpoint with a fixed timeout and the iterator will expire after the timeout even if it's still going (rather than heartbeat the checkpoint to keep the iterator alive)?

Reply:
I thought perhaps long-lived queries would not be the norm, so it might be overkill to actively refresh. I was also a little concerned about leaking checkpoints through the iterator. Perhaps we just need to expose an overall time bound in the scan options or system configuration? Then we could heartbeat as needed while still ensuring resources eventually can be reclaimed.

Reply:
I'm inclined to KISS for now and just pick some sane default expiry with no heartbeating. We can add custom expiry in ScanOptions or heartbeating later if we really need them.